src/topp/QualityControl.cpp

```diff
@@ -0,0 +1,155 @@
+  {
+  }
+protected:
+  // this function will be used to register the tool parameters
+  // it gets automatically called on tool execution
+  void registerOptionsAndFlags_() override
+  {
+    registerInputFileList_("in_raw","<file>",ListUtils::create<String>(""),"MzML input", false);
+    setValidFormats_("in_raw", ListUtils::create<String>("mzML"));
+    registerInputFileList_("in_postFDR","<file>",ListUtils::create<String>(""),"featureXML input", false);
+    setValidFormats_("in_postFDR", ListUtils::create<String>("featureXML"));
+    registerInputFile_("in_con","<file>","","Contaminant database input", false);
+    setValidFormats_("in_con",ListUtils::create<String>("fasta"));
+    //possible additions:
+    //"mzData,mzXML,dta,dta2d,mgf,featureXML,consensusXML,idXML,pepXML,fid,mzid,trafoXML,fasta"
+
+  }
+  // the main_ function is called after all parameters are read
+  ExitCodes main_(int, const char **) override
+  {
+    //-------------------------------------------------------------
+    // parsing parameters
+    //-------------------------------------------------------------
+    //
+    StringList in_raw = getStringList_("in_raw");
+    StringList in_postFDR = getStringList_("in_postFDR");
+    // Check for same length and get that length
+    QCBase::Status status;
+    UInt64 number_exps(0);
+    updateStatus_(status, number_exps, in_raw, "in_raw", QCBase::Requires::RAWMZML);
+    updateStatus_(status, number_exps, in_postFDR, "in_postFDR", QCBase::Requires::POSTFDRFEAT);
+
+    // load databases and other single file inputs
+    String in_con = getStringOption_("in_con");
+    FASTAFile fasta_file;
+    vector<FASTAFile::FASTAEntry> contaminants;
+    if (!in_con.empty())
+    {
+      fasta_file.load(in_con,contaminants);
+      status &= QCBase::Requires::CONTAMINANTS;
+    }
+
+    // Loop through file lists
+    for (Size i = 0; i < number_exps; ++i)
+    {
+      //-------------------------------------------------------------
+      // reading input
+      //-------------------------------------------------------------
+      MzMLFile mzml_file;
+      PeakMap exp;
+      if (!in_raw.empty())
+      {
+        mzml_file.load(in_raw[i],exp);
+      }
+
+      FeatureXMLFile fxml_file;
+      FeatureMap fmap;
+      if (!in_postFDR.empty())
+      {
+        fxml_file.load(in_postFDR[i], fmap);
+      }
+      //-------------------------------------------------------------
+      // calculations
+      //-------------------------------------------------------------
+    }
+    //-------------------------------------------------------------
+    // writing output
+    //-------------------------------------------------------------
+    return EXECUTION_OK;
+  }
+
+private:
+  void updateStatus_(QCBase::Status& status, UInt64& number_exps, const StringList& files, const String& port, const QCBase::Requires& req)
+  {
+    if (!files.empty())
```

Comment:
document a little?
// since files are optional, leave when not provided by user

...

Reply:
suggestion: inverse conditon and be done (no block)
`if (files.empty) return;`